plugins/ingest/src/main/java/org/elasticsearch/ingest/Data.java

```diff
@@ -42,14 +45,34 @@ public Data(String index, String type, String id, Map<String, Object> document)
         this.document = document;
     }
 
+    // TODO(talevy): support elements of lists
     @SuppressWarnings("unchecked")
     public <T> T getProperty(String path) {
         return (T) XContentMapValues.extractValue(path, document);
     }
 
-    public void addField(String field, Object value) {
+    /**
+     * add `value` to path in document. If path does not exist,
+     * nested hashmaps will be put in as parent key values until
+     * leaf key name in path is reached.
+     *
+     * @param path The path within the document in dot-notation
+     * @param value The value to put in for the path key
+     */
+    public void addField(String path, Object value) {
         modified = true;
-        document.put(field, value);
+
+        String[] pathElements = Strings.splitStringToArray(path, '.');
+
+        String writeKey = pathElements[pathElements.length - 1];
+        Map<String, Object> inner = document;
+
+        for (int i = 0; i < pathElements.length - 1; i++) {
+            inner.putIfAbsent(pathElements[i], new HashMap<String, Object>());
```

Comment:
When we back port this to the 2.x branch then this will fail because putIfAbsent is java8 only... Fixing this would be easy and I think in master we should keep this. Maybe just add a comment saying that when we back port we need to add a helper method or something like that.